test/Test/Pos/Ssc/GodTossing/Toss/BaseSpec.hs

```diff
@@ -0,0 +1,573 @@
+        richmenWithShares <- sublistOf richmenWithCerts
+        let n = length richmenWithShares
+
+        stableCerts <- HM.fromList <$>
+            mapM (\r -> (,) <$> pure r <*> (arbitrary :: Gen VssCertificate))
+                 richmenWithCerts
+
+        -- Like with 'GoodCommsPayload', only the simplest case is considered w.r.t. the
+        -- 'lastKnownEoS' field: 'checkSharesPayload' is called with the same epoch as the
+        -- 'lastKnownEoS' in 'gsVssCertificate'.
+        _gsVssCertificates <- VssCertData
+            <$> (pure . EpochOrSlot . Right . crucialSlot $ epoch)
+            <*> pure stableCerts
+            <*> arbitrary
+            <*> arbitrary
+            <*> arbitrary
+            <*> arbitrary
+
+        innerMaps <- (vector n) :: Gen [InnerSharesMap]
+        let sharesMap = HM.fromList $ zip richmenWithShares innerMaps
+            -- Every key from every inner shares map must have a commitment in the global
+            -- state i.e. be a key in '_gsCommitments'.
+            necessaryKeys = concatMap HM.keys innerMaps
+
+        _gsCommitments <- mkCommitmentsMapUnsafe <$> do
+            necessaryMap <- HM.fromList <$>
+                mapM (\k -> (,) <$> pure k <*> arbitrary) necessaryKeys
+            fillerMap <- arbitrary :: Gen (HashMap StakeholderId SignedCommitment)
+            return $ HM.union necessaryMap fillerMap
+        -- The keys in the shares map with which 'checkSharesPayload' is ran must not
+        -- exist in '_gsShares' for the global data to be correct.
+        _gsShares <-
+            customHashMapGen (arbitrary `suchThat` (not . flip elem richmenWithShares))
+                             arbitrary
+
+        return (epoch, GtGlobalState {..}, sharesMap, mrs)
+
+-- NOTE: this test does not care for 'DecrSharesNotMatchCommitment' failure. This would
+--make the already non-trivial arbitrary instance for 'GoodSharesPayload' unmanageable.
+checksGoodSharesPayload :: GoodSharesPayload -> Bool
+checksGoodSharesPayload (getGoodShares -> (epoch, gtgs, sharesMap, mrs)) =
+    case tossRunner mrs gtgs $ checkSharesPayload epoch sharesMap of
+        Left (DecrSharesNotMatchCommitment _) -> True
+        Right _ -> True
+        _ -> False
+
+-- | Checks that when the data 'checkSharesPayload' is passed is incorrect w.r.t.
+-- 'checkSharesPayload', said function fails.
+-- NOTE: does not check for 'DecrSharesNotMatchCommitment' failure. This would make the
+-- already non-trivial arbitrary instance for 'GoodSharesPayload' unmanageable.
+checksBadSharesPayload
+    :: GoodSharesPayload
+    -> StakeholderId
+    -> NonEmpty (AsBinary Share)
+    -> VssCertificate
+    -> Property
+checksBadSharesPayload
+    (getGoodShares -> (epoch, g@GtGlobalState {..}, sm, mrs))
+    sid
+    ne
+    cert =
+    let -- This property assumes the existence of a stakeholder not in the commitments or
+        -- shares map. Instead of writing a new 'Arbitrary' type which will only be useful
```

Comment:
> the existence of a stakeholder not in the commitments or shares map

It sounds strange to me, but maybe that's because I am not a native speaker. Perhaps it can be reformulated more clearly using word `absence` :wink: